mne/time_frequency/tfr.py

```diff
@@ -3105,8 +3105,14 @@ def __setstate__(self, state):
         ).squeeze(axis=0)
         self.events = state.get("events", _ensure_events(fake_events))
         self.event_id = state.get("event_id", _check_event_id(None, self.events))
-        self.drop_log = state.get("drop_log", tuple())
         self.selection = state.get("selection", np.arange(n_epochs))
+        self.drop_log = state.get(
+            "drop_log",
+            tuple(
+                () if k in self.selection else ("IGNORED",)
+                for k in range(max(len(self.events), max(self.selection) + 1))
```

Comment:
scratching my head about this line. When would we *not* want just `range(len(self.events))`?

Reply:
We always have `len(self.events)` == `len(epochs)`. But let's say you originally had 100 events and just selected the last 10 on way or another, that resulting `Epochs` object shouldhave for example
```
len(epochs) == 10
len(epochs.selection) == 10
len(epochs.events) == 10
list(epochs.selection) == list(range(90, 100))
```
and in general for anything not in `epochs.selection` you should be able to query that index in `epochs.drop_log` to see why it's not part of `epochs`. So `epochs.drop_log[30]` for example should exist and be non-empty meaning it was dropped for a reason, whereas `epochs.drop_log[90:]` should exist and all be empty (i.e., those are the epochs that have been kept).

Reply:
right, I forgot about events getting truncated/sliced/subselected.

Am I right then that we don't need to care about the case of having 30 events, and selecting only the *middle* 10?  IIUC, if that happened via normal means, the `drop_log` would be present and the code in this diff won't be reached. It's only for EpochsTFRArray (where we don't know the dropping history) where we have to spawn a "fake" drop log, and in that case we can't know if (say) the last 10 epochs were dropped, but we also don't *need* to know that for things to work correctly.

Reply:
Yeah I think we just need some drop log that satisfies the necessary expectations about length of the drop log, selection values, and number of events.

Reply:
So something like
```
tuple(() if k in self.selection else ("IGNORED",) for k in len(self.events))
```
to account for the fact that a non-default `selection` param could be passed?

Reply:
No the `len(self.events)` won't be enough, I think what you have here with the `max(...)` is more correct

Reply:
... or you have to make sure `self.selection = np.arange(len(self.events))`

Reply:
Ah sorry. Yeah at the moment whatever selection can be passed so `self.selection = np.arange(len(self.events))` can't be assumed.